pkg/cloudprovider/providers/vsphere/nodemanager.go

```diff
@@ -50,117 +50,142 @@ type NodeManager struct {
-
-		type VmSearchResult struct {
-			vm *vclib.VirtualMachine
-			datacenter *vclib.Datacenter
-			err error
-		}
 
-		c1 := make(chan *VmSearchResult)
-		nodeUUID := node.Status.NodeInfo.SystemUUID
-		for _, datacenterObj := range datacenterObjs {
-			glog.V(4).Infof("Finding node %s in vc=%s and datacenter=%s", node.Name, vc, datacenterObj.Name())
-			go func(dc *vclib.Datacenter) {
-				vm, err := dc.GetVMByUUID(ctx, nodeUUID)
-				c1 <- &VmSearchResult{
-					vm:         vm,
-					err:        err,
+			for _, datacenterObj := range datacenterObjs {
+				glog.V(4).Infof("Finding node %s in vc=%s and datacenter=%s", node.Name, vc, datacenterObj.Name())
+				queueChannel <- &VmSearch{
+					vc:         vc,
 					datacenter: datacenterObj,
 				}
-			}(datacenterObj)
-		}
-
-		for i := 0; i < len(datacenterObjs); i++ {
-			res := <-c1
-			if res.vm != nil {
-				glog.V(4).Infof("Found node %s as vm=%+v in vc=%s and datacenter=%s",
-					node.Name, res.vm, vc, res.datacenter.Name())
-
-				nodeInfo := &NodeInfo{dataCenter: datacenterObjs[i], vm: res.vm, vcServer: vc}
-				nm.nodeInfoLock.Lock()
-				nm.nodeInfoMap[node.ObjectMeta.Name] = nodeInfo
-				nm.nodeInfoLock.Unlock()
-			} else {
-				glog.V(4).Infof("Did not find node %s in vc=%s and datacenter=%s",
-					node.Name, vc, res.datacenter.Name())
 			}
 		}
+		close(queueChannel)
+	}()
+
+	for i := 0; i < POOL_SIZE; i++ {
+		go func() {
+			for res:= range queueChannel {
+				ctx, cancel := context.WithCancel(context.Background())
+				defer cancel()
+				vm, err := res.datacenter.GetVMByUUID(ctx, nodeUUID)
+				if vm != nil {
+					glog.V(4).Infof("Found node %s as vm=%+v in vc=%s and datacenter=%s",
+						node.Name, vm, res.vc, res.datacenter.Name())
+
+					nodeInfo := &NodeInfo{dataCenter: res.datacenter, vm: vm, vcServer: res.vc}
+					nm.SetNodeInfo(node.ObjectMeta.Name, nodeInfo)
+					for range queueChannel {}
+					break
+				} else {
+					glog.V(4).Infof("Did not find node %s in vc=%s and datacenter=%s",
+						node.Name, res.vc, res.datacenter.Name(), err)
+				}
+			}
+			wg.Done()
+		}()
+		wg.Add(1)
 	}
-
-	return nil
+	wg.Wait()
 }
 
 
+
 func (nm *NodeManager) RegisterNode(node *v1.Node) error {
-	nm.registeredNodesLock.Lock()
-	nm.registeredNodes[node.ObjectMeta.Name] = node
-	nm.registeredNodesLock.Unlock()
+	nm.SetNode(node)
 	nm.DiscoverNode(node)
 	return nil
 }
 
 func (nm *NodeManager) UnRegisterNode(node *v1.Node) error {
-	nm.registeredNodesLock.Lock()
-	delete(nm.registeredNodes, node.ObjectMeta.Name)
-	nm.registeredNodesLock.Unlock()
-
+	nm.RemoveNode(node)
 	return nil
 }
 
 func (nm *NodeManager) RediscoverNode(nodeName k8stypes.NodeName) error {
-	nm.registeredNodesLock.RLock()
-	node := nm.registeredNodes[convertToString(nodeName)]
-	nm.registeredNodesLock.RUnlock()
+	node, err := nm.GetNode(nodeName)
 
-	if node == nil {
-		return ErrNoNodeInfo
+	if err != nil {
+		return err
 	}
 	nm.DiscoverNode(node)
 	return nil
```

Comment:
RediscoverNode should return error if discovery fails. This way, the caller will know if the rediscovery worked or failed.